### PR TITLE
optimize replication padding performance on CPU

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102949
* #102597
* #102518
* __->__ #102255
* #102254

The major difference from the previous PR on ReflectionPad is the padding indexing struct, `ReplicationPad::index()`, the rest of the part is pretty much the same.

The following benchmark result gathered on Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz, with 20 cores per socket.

### single core inference
```
(before)
ReplicationPad2d((2, 2, 2, 2)) size:  torch.Size([1, 3, 224, 224]) , NCHW: 0.265 ms; 
ReplicationPad2d((2, 2, 2, 2)) size:  torch.Size([128, 64, 56, 56]) , NCHW: 52.336 ms;

(after)
ReplicationPad2d((2, 2, 2, 2)) size:  torch.Size([1, 3, 224, 224]) , NCHW: 0.048 ms;  
ReplicationPad2d((2, 2, 2, 2)) size:  torch.Size([128, 64, 56, 56]) , NCHW: 17.199 ms;  
```

### single socket inference
```
(before)
ReplicationPad2d((2, 2, 2, 2)) size:  torch.Size([1, 3, 224, 224]) , NCHW: 0.111 ms; 
ReplicationPad2d((2, 2, 2, 2)) size:  torch.Size([128, 64, 56, 56]) , NCHW: 3.885 ms;  

(after)
ReplicationPad2d((2, 2, 2, 2)) size:  torch.Size([1, 3, 224, 224]) , NCHW: 0.011 ms; 
ReplicationPad2d((2, 2, 2, 2)) size:  torch.Size([128, 64, 56, 56]) , NCHW: 3.148 ms;  
```

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 1]
```diff
@@ -973,72 +314,18 @@ Tensor replication_pad2d_backward_cpu(
...
-  }
+  replication_pad3d_kernel(kCPU, output, input, paddingSize);
 }
 
 Tensor& replication_pad3d_backward_out_cpu(const Tensor& gradOutput,
     const Tensor& input,
     IntArrayRef paddingSize,
     Tensor& gradInput)
 {
+  gradInput.resize_as_(input);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/ReplicationPadding.cpp:327]: Why is this change needed? Did the previous PR in this stack introduce a bag that is now being fixed in this one?

Comment (mingfeima) [reply aten/src/ATen/native/ReplicationPadding.cpp:327]: This has a signature using `grad_input` as an alias:
```
replication_pad3d_backward.grad_input(Tensor grad_output, Tensor self, SymInt[6] padding, *, Tensor(a!) grad_input) -> Tensor(a!)
```

Originally, the reflection padding has this `grad_input.resize_as_(input)` but replication padding doesn't have this.
The resize logic is needed. IF the user manage to call the `replication_pad3d_backward(grad_output, self, padding, grad_input)` and `grad_input` is given with a wrong shape, it breaks. This should be a **very weird usage** but i think it should be better to add a resize logic when `grad_input` is given as an alias.

**Review by cpuhrsch — APPROVED** [review]

File: aten/src/ATen/native/ReplicationPadding.cpp (hunk 1 above)
Comment (cpuhrsch) [reply aten/src/ATen/native/ReplicationPadding.cpp:327]: Ok, thank you for double checking and the explanation!
